### PR TITLE
replace prints by hirte_log calls

### LINKED ISSUE #96: Better error handling and logging framework
Our current handling of errors is rather naive. All we do is print stuff on stderr and return false. Once hirte is in production I think we want something more manageable. For example, we should probably have some minimal helpers for logging issues that have a log level which can be tweaked during testing via the configuration. And, it should probably go to the systemd journal at least when in production (although during testing, being able to get it to stderr is probably nice too).

We also shoud probably pick up the systemd error reporting mechanism of returning `-errno` style integers instead of just true/false for "success"/"error". That way callers can get at least some more nformation on exactly what failed. In many cases we already call into systemd APIs which can give this info. For example:

```
bool node_export(Node *node) {
        Manager *manager = node->manager;

        int r = sd_bus_add_object_vtable(
                        manager->user_dbus,
                        &node->export_slot,
                        node->object_path,
                        NODE_INTERFACE,
                        node_vtable,
                        node);
        if (r < 0) {
                fprintf(stderr, "Failed to add node vtable: %s\n", strerror(-r));
                return false;
        }

        return true;
}
```

This would probably be nicer if it was something like:

```
int node_export(Node *node) {
        Manager *manager = node->manager;

        int r = sd_bus_add_object_vtable(
                        manager->user_dbus,
                        &node->export_slot,
                        node->object_path,
                        NODE_INTERFACE,
                        node_vtable,
                        node);
        if (r < 0) {
                hirte_log(HIRTE_LOG_WARNING, "Failed to add node vtable: %s\n", strerror(-r));
        }

        return r;
}
```

### PR DESCRIPTION
**Pull request by engelmi**
This PR 
- exchanges all `fprintf`s etc. with the new `hirte_log_*` calls
- lets the logging be configured by the config ini
- does a bit of cleanup

### CODE REVIEW

**Comment by rhatdan** [comment]
LGTM

**Comment by alexlarsson** [comment]
This conflicted with the properties/monitors PR, and also, that added a lot of new prints that need to be converted :)

**Comment by engelmi** [comment]
> This conflicted with the properties/monitors PR, and also, that added a lot of new prints that need to be converted :)

Yes, I was expecting a few conflicts. Rebased and exchanged the prints with the logging functions :)

File: src/agent/agent.c [hunk 1]
```diff
@@ -886,8 +889,6 @@ bool agent_start(Agent *agent) {
         sd_bus_error error = SD_BUS_ERROR_NULL;
         _cleanup_sd_bus_message_ sd_bus_message *m = NULL;
 
-        fprintf(stdout, "Starting Agent...\n");
```
Comment (pypingou) [inline src/agent/agent.c:889]: You don't want to keep this one? (If we do, we may want to specify: `Start hirte agent`)

File: src/agent/agent.c [hunk 2]
```diff
@@ -1047,43 +1048,41 @@ bool agent_start(Agent *agent) {
...
-                fprintf(stderr, "Starting event loop failed: %s\n", strerror(-r));
+                hirte_log_errorf("Starting event loop failed: %s", strerror(-r));
                 return false;
         }
 
         return true;
 }
 
 bool agent_stop(Agent *agent) {
-        fprintf(stdout, "Stopping Agent...\n");
```
Comment (pypingou) [inline src/agent/agent.c:1085]: same as for the starting above

File: src/agent/main.c [hunk 3]
```diff
@@ -96,7 +101,5 @@ int main(int argc, char *argv[]) {
                 return EXIT_SUCCESS;
         }
 
-        fprintf(stdout, "Agent exited\n");
```
Comment (pypingou) [inline src/agent/main.c:99]: Same as above :)

File: src/client/client.c [hunk 4]
```diff
@@ -1,9 +1,10 @@
-#include "client.h"
 #include "libhirte/bus/utils.h"
-#include <stdio.h>
+#include "libhirte/log/log.h"
+
+#include "client.h"
 
 int list_units_on_all(UNUSED Client *client) {
-        printf("Listing units on all\n");
+        hirte_log_info("Listing units on all");
```
Comment (pypingou) [inline src/client/client.c]: hm, are we sure about this? This is the client, so we want it to print things to stdout, not log them somewhere

File: src/client/main.c [hunk 5]
```diff
@@ -17,7 +18,7 @@ static int opargc;
 static bool no_action = false;
 
 static void usage(char *argv[]) {
-        fprintf(stderr, "Usage: %s COMMAND ... \n", argv[0]);
+        hirte_log_errorf("Usage: %s COMMAND ... ", argv[0]);
```
Comment (pypingou) [inline src/client/main.c]: same as for client.c above, we want this to be user-visible

**Review by pypingou — COMMENTED** [review]
Overall looks good to me, but I'm not sure (because I don't know enough how it works), if we really want to replace the printf and fprintf calls for the client.
For the client, we would want most of its output to stdout and not stderr which is where I expect we'll be logging most of our other messages (which the example config files also confirm)

File: src/agent/main.c [hunk 6]
```diff
@@ -24,7 +24,7 @@ static const char *opt_config = NULL;
 // NOLINTEND(cppcoreguidelines-avoid-non-const-global-variables)
 
 static void usage(char *argv[]) {
-        fprintf(stderr, "Usage: %s [-H host] [-p port] [-c config] [-n name]\n", argv[0]);
+        hirte_log_errorf("Usage: %s [-H host] [-p port] [-c config] [-n name]\n", argv[0]);
```
Comment (alexlarsson) [inline src/agent/main.c]: I don't think usage() should be using logging. This is a straight print.

File: src/agent/main.c [hunk 7]
```diff
@@ -54,18 +54,20 @@ static void get_opts(int argc, char *argv[]) {
                         break;
 
                 default:
-                        fprintf(stderr, "Unsupported option %c\n", opt);
+                        hirte_log_errorf("Unsupported option %c\n", opt);
```
Comment (alexlarsson) [inline src/agent/main.c]: Same here, option parsing errors is imho straight stderr prints too.

File: src/client/main.c [hunk 8]
```diff
@@ -17,7 +18,7 @@ static int opargc;
 static bool no_action = false;
 
 static void usage(char *argv[]) {
-        fprintf(stderr, "Usage: %s COMMAND ... \n", argv[0]);
+        hirte_log_errorf("Usage: %s COMMAND ... \n", argv[0]);
```
Comment (alexlarsson) [inline src/client/main.c]: same

Comment (alexlarsson) [reply src/client/main.c]: and below for errors

File: src/manager/main.c [hunk 9]
```diff
@@ -21,7 +22,7 @@ static const char *opt_config = NULL;
 // NOLINTEND(cppcoreguidelines-avoid-non-const-global-variables)
 
 static void usage(char *argv[]) {
-        fprintf(stderr, "Usage: %s [-p port] [-c config]\n", argv[0]);
+        hirte_log_errorf("Usage: %s [-p port] [-c config]\n", argv[0]);
```
Comment (alexlarsson) [inline src/manager/main.c]: again

File: src/libhirte/log/log.c [hunk 10]
```diff
@@ -87,6 +110,41 @@ void hirte_log_init() {
...
+
+        target = topic_lookup(logging_topic, "Target");
+        hirte_log_set_log_fn(hirte_log_to_stderr_with_location);
+        if (target && streq("journald", target)) {
+                hirte_log_set_log_fn(hirte_log_to_journald_with_location);
+        }
+
+        quiet = topic_lookup(logging_topic, "Quiet");
+        hirte_log_set_quiet(false);
+        if (quiet && streq("true", quiet)) {
```
Comment (alexlarsson) [inline src/libhirte/log/log.c]: I would prefer a config helper function here that is a bit more relaxed to what parses as true. Accepting maybe also "yes", "on", and "t" in a case insensitive way,

File: src/agent/agent.c (hunk 1 above)
Comment (engelmi) [reply src/agent/agent.c:889]: This was one of the very first dev debug prints I added. Not sure if it provides much value. We probably need to use `hirte` a bit more and get a feeling where the logging is lacking and need to adjust it. For now, I'd remove it.

File: src/agent/agent.c (hunk 2 above)
Comment (engelmi) [reply src/agent/agent.c:1085]: Same as [my previous reply](https://github.com/containers/hirte/pull/116#discussion_r1101199348)

File: src/agent/main.c (hunk 3 above)
Comment (engelmi) [reply src/agent/main.c:99]: Same as [my previous reply](https://github.com/containers/hirte/pull/116#discussion_r1101199348)

File: src/client/client.c (hunk 4 above)
Comment (engelmi) [reply src/client/client.c]: Good point! I removed the logging from the client for now.

File: src/client/main.c (hunk 5 above)
Comment (engelmi) [reply src/client/main.c]: Changed :)

File: src/agent/main.c (hunk 6 above)
Comment (engelmi) [reply src/agent/main.c]: Yes, you are right. Changed/reverted it.

File: src/agent/main.c (hunk 7 above)
Comment (engelmi) [reply src/agent/main.c]: Changed.

File: src/client/main.c (hunk 8 above)
Comment (engelmi) [reply src/client/main.c]: Changed.

File: src/manager/main.c (hunk 9 above)
Comment (engelmi) [reply src/manager/main.c]: Changed.

File: src/client/client.c [hunk 11]
```diff
@@ -1,9 +1,9 @@
-#include "client.h"
 #include "libhirte/bus/utils.h"
-#include <stdio.h>
+
+#include "client.h"
 
 int list_units_on_all(UNUSED Client *client) {
-        printf("Listing units on all\n");
+        printf("Listing units on all");
```
Comment (pypingou) [inline src/client/client.c]: missing the trailing "\n" ;)

(and on a few below, cf the diff here)

File: src/client/main.c [hunk 12]
```diff
@@ -17,7 +17,7 @@ static int opargc;
 static bool no_action = false;
 
 static void usage(char *argv[]) {
-        fprintf(stderr, "Usage: %s COMMAND ... \n", argv[0]);
+        printf("Usage: %s COMMAND ... ", argv[0]);
```
Comment (pypingou) [inline src/client/main.c]: same about the trailing "\n" (unless that's the difference between printf and fprintf)

File: src/agent/agent.c (hunk 1 above)
Comment (pypingou) [reply src/agent/agent.c:889]: if you do `systemctl start hirte`, I think it's nice to have in `journalctl` the entry `Starting hirte-agent`, it helps auditability later on

File: src/client/main.c (hunk 12 above)
Comment (engelmi) [reply src/client/main.c]: Yes, its missing... damn those simple C prints!

File: src/agent/agent.c (hunk 1 above)
Comment (alexlarsson) [reply src/agent/agent.c:889]: systemd would log the start though.

File: src/client/client.c (hunk 11 above)
Comment (engelmi) [reply src/client/client.c]: Changed.

File: src/libhirte/log/log.c (hunk 10 above)
Comment (engelmi) [reply src/libhirte/log/log.c]: Good idea! Please have another look @alexlarsson I implemented the `is_true_value` function using `strcasecmp`.

File: src/client/main.c [hunk 13]
```diff
@@ -29,7 +29,7 @@ static int get_opts(int argc, char *argv[]) {
                         no_action = true;
                         return 0;
                 } else {
-                        fprintf(stderr, "Unsupported option %c\n", opt);
+                        fprintf(stderr, "Unsupported option %c", opt);
```
Comment (pypingou) [inline src/client/main.c]: Missed one here ;)

Comment (engelmi) [reply src/client/main.c]: Added.

**Review by alexlarsson — APPROVED** [review]

**Pull request merged**